graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/functions/messages/CloneMessage.java

```diff
@@ -26,26 +28,56 @@
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
+import java.util.Objects;
+import java.util.Optional;
+
 import static org.graylog.plugins.pipelineprocessor.ast.functions.ParameterDescriptor.type;
 
 public class CloneMessage extends AbstractFunction<Message> {
     private static final Logger LOG = LoggerFactory.getLogger(CloneMessage.class);
 
     public static final String NAME = "clone_message";
 
+    private static final String CLONE_SOURCE = "gl2_clone_source_rule";
+    private static final String CLONE_NUMBER = "gl2_clone_number";
+    static final int MAX_CLONES = 100;
+
     private final ParameterDescriptor<Message, Message> messageParam;
+    private final ParameterDescriptor<Boolean, Boolean> loopDetectionParam;
 
     public CloneMessage() {
+        loopDetectionParam = ParameterDescriptor.bool("preventLoops").optional().description("Detects if a cloned message is processed by the same rule again, in order to prevent loops. Defaults to 'false', but will not allow more than " + MAX_CLONES + " clones if not explicitly set.").build();
         messageParam = type("message", Message.class).optional().description("The message to use, defaults to '$message'").build();
     }
 
     @Override
     public Message evaluate(FunctionArgs args, EvaluationContext context) {
         final Message currentMessage = messageParam.optional(args, context).orElse(context.currentMessage());
+        final Optional<Boolean> preventLoops = loopDetectionParam.optional(args, context);
+
+        int cloneNumber = ObjectUtils.defaultIfNull(currentMessage.getFieldAs(Integer.class, CLONE_NUMBER), 0);
+
+        final Rule cloneSource = currentMessage.getFieldAs(Rule.class, CLONE_SOURCE);
+        final Rule rule = context.getRule();
+        if (ObjectUtils.allNotNull(cloneSource, rule) && Objects.equals(cloneSource, rule)) { // message was cloned by the same rule before
+
+            if (preventLoops.orElse(false)) {
+                return null;
+            }
+
+            if (preventLoops.isEmpty() && cloneNumber >= MAX_CLONES) {
+                LOG.error("Message was cloned more than {} times by rule '{}'. Not allowing any more clones to prevent a potential endless loop. If this was intentional, please explicitly set the 'preventLoops' parameter to 'false'.",
+                        MAX_CLONES, cloneSource.name());
+                return null;
+            }
+
+        }
 
         final Message clonedMessage = new Message(currentMessage.getMessage(), currentMessage.getSource(), currentMessage.getTimestamp());
         clonedMessage.addFields(currentMessage.getFields());
         clonedMessage.addStreams(currentMessage.getStreams());
+        clonedMessage.addField(CLONE_SOURCE, rule);
+        clonedMessage.addField(CLONE_NUMBER, ++cloneNumber);
```

Comment:
How about using the metadata fields in `Message` instead of setting actual fields in the message?

https://github.com/Graylog2/graylog2-server/blob/ed3a4da30364e786df619a3067c1854346752791/graylog2-server/src/main/java/org/graylog2/plugin/Message.java#L1011-L1064

Otherwise, those fields would end up being indexed in OpenSearch. Or do I miss the removal of the fields somewhere?

Reply:
thanks for the comment. I'll change that.